File: backend/app/routers/tickets.py
import uuid
from fastapi.requests import Request
from pydantic import BaseModel
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import require_api_key, require_clerk_agent
from app.models.magic_token import MagicToken
from app.models.ticket import Ticket, TicketStatus
from app.schemas.ticket import AssignPayload, ReplyPayload, TicketCreate, TicketOut
from app.models.agent import Agent
from app.schemas.ticket import TicketStatusOut
from app.middleware.rate_limit import TICKET_CREATE_LIMIT, limiter

router = APIRouter(prefix="/tickets", tags=["tickets"])

@router.post("", response_model=TicketOut, status_code=201)
@limiter.limit(TICKET_CREATE_LIMIT)
def create_ticket(
    request: Request,
    body: TicketCreate,
    db: Session = Depends(get_db),
    _: bool = Depends(require_api_key),
):
    ticket = Ticket(
        subject=body.subject,
        body=body.body,
        customer_email=body.customer_email,
        customer_name=body.customer_name,
    )
    db.add(ticket)
    db.flush()  # get ticket.id before magic token insert

    magic_token = MagicToken(ticket_id=ticket.id, token=str(uuid.uuid4()))
    db.add(magic_token)
    db.commit()
    db.refresh(ticket)
    return ticket


class TicketListOut:
    pass

class TicketPage(BaseModel):
    items: List[TicketOut]
    total: int


@router.get("", response_model=TicketPage)
def list_tickets(
    status: Optional[TicketStatus] = Query(None),
    priority: Optional[str] = Query(None),
    category: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
    _agent: dict = Depends(require_clerk_agent),
):
    q = db.query(Ticket)
    if status:
        q = q.filter(Ticket.status == status)
    if priority:
        q = q.filter(Ticket.priority == priority)
    if category:
        q = q.filter(Ticket.category == category)

    total = q.count()
    items = (
        q.order_by(Ticket.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )
    return TicketPage(items=items, total=total)


@router.get("/{ticket_id}", response_model=TicketOut)
def get_ticket(
    ticket_id: str,
    db: Session = Depends(get_db),
    _agent: dict = Depends(require_clerk_agent),
):
    ticket = db.query(Ticket).filter(Ticket.id == ticket_id).first()
    if ticket is None:
        raise HTTPException(status_code=404, detail="Ticket not found")
    return ticket


@router.patch("/{ticket_id}/assign", response_model=TicketOut)
def assign_ticket(
    ticket_id: str,
    body: AssignPayload,
    db: Session = Depends(get_db),
    _agent: dict = Depends(require_clerk_agent),
):
    ticket = db.query(Ticket).filter(Ticket.id == ticket_id).first()
    if ticket is None:
        raise HTTPException(status_code=404, detail="Ticket not found")

    agent = db.query(Agent).filter(Agent.id == body.agent_id).first()
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")

    ticket.assigned_agent_id = body.agent_id
    db.commit()
    db.refresh(ticket)
    return ticket


@router.patch("/{ticket_id}/reply", response_model=TicketOut)
def save_reply(
    ticket_id: str,
    body: ReplyPayload,
    db: Session = Depends(get_db),
    _agent: dict = Depends(require_clerk_agent),
):
    ticket = db.query(Ticket).filter(Ticket.id == ticket_id).first()
    if ticket is None:
        raise HTTPException(status_code=404, detail="Ticket not found")

    ticket.agent_reply = body.agent_reply
    ticket.status = TicketStatus.in_progress
    db.commit()
    db.refresh(ticket)
    return ticket


@router.patch("/{ticket_id}/resolve", response_model=TicketOut)
def resolve_ticket(
    ticket_id: str,
    db: Session = Depends(get_db),
    _agent: dict = Depends(require_clerk_agent),
):
    ticket = db.query(Ticket).filter(Ticket.id == ticket_id).first()
    if ticket is None:
        raise HTTPException(status_code=404, detail="Ticket not found")

    ticket.status = TicketStatus.resolved
    db.commit()
    db.refresh(ticket)
    # TODO:push outbound email to qstash
    return ticket


@router.get("/{ticket_id}/status", response_model=TicketStatusOut)
def get_ticket_status(
    ticket_id: str,
    token: str = Query(...),
    db: Session = Depends(get_db),
):
    from app.dependencies import require_magic_token

    magic = require_magic_token(token=token, db=db)
    if magic.ticket_id != ticket_id:
        raise HTTPException(status_code=404, detail="Token does not match ticket")

    ticket = db.query(Ticket).filter(Ticket.id == ticket_id).first()
    if ticket is None:
        raise HTTPException(status_code=404, detail="Ticket not found")
    return ticket

File: frontend/src/shared/api/tickets.ts
import { API_KEY, API_URL, apiFetch } from "./client";

export interface Ticket {
  id: string;
  subject: string;
  body: string;
  source: "email" | "form";
  status: "open" | "in_progress" | "resolved";
  priority: "low" | "medium" | "high";
  category: "billing" | "technical" | "general" | "other";
  escalate: boolean;
  ai_draft_reply: string | null;
  agent_reply: string | null;
  assigned_agent_id: string | null;
  customer_email: string | null;
  customer_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface TicketPage {
  items: Ticket[];
  total: number;
}

export interface TicketStatusResponse {
  id: string;
  subject: string;
  status: "open" | "in_progress" | "resolved";
  agent_reply: string | null;
  created_at: string;
  updated_at: string;
}

export interface Agent {
  id: string;
  email: string;
  name: string;
}

export interface TicketCreatePayload {
  subject: string;
  body: string;
  customer_email: string;
  customer_name: string;
}

export interface TicketCreateResult extends Ticket {
  magic_token: string;
}

export function createTicket(
  payload: TicketCreatePayload,
  signal?: AbortSignal,
): Promise<TicketCreateResult> {
  return apiFetch<TicketCreateResult>(`${API_URL}/tickets`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "x-api-key": API_KEY },
    body: JSON.stringify(payload),
    signal,
  });
}

export function listTickets({
  token,
  status,
  priority,
  category,
  page = 1,
  signal,
}: {
  token: string;
  status?: string;
  priority?: string;
  category?: string;
  page?: number;
  signal?: AbortSignal;
}): Promise<TicketPage> {
  const params = new URLSearchParams();
  if (status) params.set("status", status);
  if (priority) params.set("priority", priority);
  if (category) params.set("category", category);
  params.set("page", String(page));
  return apiFetch<TicketPage>(`${API_URL}/tickets?${params}`, {
    headers: { Authorization: `Bearer ${token}` },
    signal,
  });
}

export function getTicket(id: string, token: string, signal?: AbortSignal): Promise<Ticket> {
  return apiFetch<Ticket>(`${API_URL}/tickets/${encodeURIComponent(id)}`, {
    headers: { Authorization: `Bearer ${token}` },
    signal,
  });
}

export function listAgents(token: string, signal?: AbortSignal): Promise<Agent[]> {
  return apiFetch<Agent[]>(`${API_URL}/agents`, {
    headers: { Authorization: `Bearer ${token}` },
    signal,
  });
}

export function assignTicket(
  id: string,
  agentId: string,
  token: string,
  signal?: AbortSignal,
): Promise<Ticket> {
  return apiFetch<Ticket>(`${API_URL}/tickets/${encodeURIComponent(id)}/assign`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
    body: JSON.stringify({ agent_id: agentId }),
    signal,
  });
}

export function saveReply(
  id: string,
  agentReply: string,
  token: string,
  signal?: AbortSignal,
): Promise<Ticket> {
  return apiFetch<Ticket>(`${API_URL}/tickets/${encodeURIComponent(id)}/reply`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
    body: JSON.stringify({ agent_reply: agentReply }),
    signal,
  });
}

export function resolveTicket(id: string, token: string, signal?: AbortSignal): Promise<Ticket> {
  return apiFetch<Ticket>(`${API_URL}/tickets/${encodeURIComponent(id)}/resolve`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
    body: JSON.stringify({}),
    signal,
  });
}

export function getTicketStatus(
  id: string,
  token: string,
  signal?: AbortSignal,
): Promise<TicketStatusResponse> {
  return apiFetch<TicketStatusResponse>(`${API_URL}/tickets/${encodeURIComponent(id)}/status`, {
    headers: { "x-magic-token": token },
    signal,
  });
}
